### ISSUE TITLE
Extend MQTT output topic formatting options 

### ISSUE THREAD

**Issue by fietser28** [influxdata/telegraf#10732]
## Feature Request

The current output options of the MQTT output plugin are limited and have some almost hardcoded assumptions (hostname). I would like to propose an extension suiting the nature of MQTT. 

This proposal is  almost 'the inverse' of #10712. 
### Proposal:

Support something like:

Input:
```
'''
sensors,site=CLE,version=v1,device_name=device5,field=temp1 value=21.0 (float)
sensors,site=CLE,version=v1,device_name=device5,field=spin2 value=100i (int)
sensors,site=CLE,version=v1,device_name=device5,field=status value="OK" (string)
...
```

New proposed type of config:
```
[[output.mqtt_producer]]
    ....
    [[output.mqtt_producer.topic_defintion]]
    topic = "/sensors/_/_/_/_"
    measurement = "/measurement/_/_/_/_"
    tags = "_/site/version/device_name/_"
    fields = "_/_/_/field"
```
Where `topic` specifies the fixed topic parts and defines the variable part the other options have to fill (once and only once).
`measurement` and `tags` map the respective tags and measurement values in the parts of the topic. `field` is replaced with the field name.
 
Will result in MQTT messages output:
```
 Topic                       Message
 /sensors/CLE/v1/device5/temp1 21.0
 /sensors/CLE/v1/device5/temp2 21.0
 /sensors/CLE/v1/device5/rpm     100
 /sensors/CLE/v1/device5/ph       7.0
 /sensors/CLE/v1/device5/spin1   1
 /sensors/CLE/v1/device5/spin2  13
 /sensors/CLE/v1/device5/status OK
 /sensors/CLE/v1/device5/log      INFO: happy
```
(Note: this example is the exact inverse of the example in #10712) A unit test would be to "output what you can consume as input".

### Current behavior:

Currently it would require a lot of seperate MQTT producers to achieve this example.

### Desired behavior:

See example.

### Use case: 

In local IOT systems based on MQTT backbone  InfluxDB is a very powerful producer of aggregated data (e.g. daily, weekly stats) you want to push this data out to MQTT as well. JSON/influxdb parsing is not always a strong point or possible on IOT devices OR you have to adjust to existing patterns on the existing MQTT installation.

**Comment by reimda** [comment]
@srebhan Could you take a look since you worked on 10712?

**Comment by srebhan** [comment]
@fietser28 I struggled a bit by you saying it is the "inverse" but what you mean is that you want a "topic producing on the output" counterpart to "topic parsing on the input". Is my understanding correct?

**Comment by fietser28** [comment]
@srebhan: Yes. 

Maybe this helps (feel free to ignore if it confuses more):

A mathematical definition:
If you think of an input/output plugins as a mathematical functions I = Fin(M) and M = Fout(I), I = an influxdb stream and M = a MQTT stream, then:
If Iin = Fin(Fout(Iout)) then Iin = Iout 
If Mout = Fout(Fin(Min) then Mout = Min 
This requires the configs specifying Fin and Fout to be complementary. 

In plain english:
You can feed the output of the output plugin into the input plugin and with the right configs you end op with the same data stream as you fed originally into the output plugin.
 
the same operation in the opposite direction

PS. English is not my native language.

**Comment by serroba** [comment]
@fietser28 https://github.com/influxdata/telegraf/pull/12224#event-8248716785 this has been merged. May not be identical to what you had in mind, but it relates

**Comment by srebhan** [comment]
The topic can be generated via
```toml
topic = 'sensors/{{ .Tag "site" }}/{{ .Tag "version" }}/{{ .Tag "field" }}
```
However, you might need to test if `csv` can generate a single value serialization...

**Comment by srebhan** [comment]
This is related to #12589 and I'm working on a fix... Keeping this open till then...
